### PR TITLE
Stop writing to the queue if nothing is reading it

### PR DESCRIPTION
**Pull request by tipabu**

### CODE REVIEW

File: collectd_plugin/collectd_plugin.py [hunk 1]
```diff
@@ -947,5 +948,7 @@ def _should_filter(values, data, ds):
     data = {}
     collectd.register_init(init)
     collectd.register_config(config, data)
-    collectd.register_write(write, data)
+    if data.get("monitor_endpoint") or (
+            data.get("carbon_host") and data.get("carbon_port")):
+        collectd.register_write(write, data)
```
Comment (clayg) [inline collectd_plugin/collectd_plugin.py:961]: did you test this on a ssnode?  Or do you want to package it and see what happens?

Comment (tipabu) [reply collectd_plugin/collectd_plugin.py:961]: Untested -- just trying to address your concern about continuing to pump data into a `Queue` with no `writer` thread to read from it.

File: collectd_plugin/collectd_plugin.py [hunk 2]
```diff
@@ -227,9 +228,17 @@ def statsd_thread(
     if statsd_forward_prefix_hostname:
         statsd_forward_prefix = socket.gethostname()
         _log("Forwarding StatsD stats with hostname prefix %r", statsd_forward_prefix)
+
+    if data.get("monitor_endpoint") or (
+            data.get("carbon_host") and data.get("carbon_port")):
+        transport = "graphite_queue"
```
Comment (charz) [inline collectd_plugin/collectd_plugin.py:234]: @tipabu just a dump question, why we still care statsd server if we don't have `monitor_endpoint` and `carbon_host`?

Comment (charz) [reply collectd_plugin/collectd_plugin.py:234]: ~~Also, is this (https://github.com/sivy/pystatsd) a correct repo that we use in this plugin?~~ NM, I figured it out.

File: collectd_plugin/collectd_plugin.py [hunk 3]
```diff
@@ -227,9 +228,17 @@ def statsd_thread(
...
+    else:
+        # force our hacked-up pystatsd Server to use TransportNop
```
Comment (clayg) [inline collectd_plugin/collectd_plugin.py:236]: what do you mean "TransportNop"

File: collectd_plugin/collectd_plugin.py (hunk 2 above)
Comment (tipabu) [reply collectd_plugin/collectd_plugin.py:234]: We still need the statsd endpoint to receive metrics from swift and forward them to the statsd-exporter, no?

File: collectd_plugin/collectd_plugin.py (hunk 3 above)
Comment (tipabu) [reply collectd_plugin/collectd_plugin.py:236]: Look for `TransportGraphiteQueue` and `TransportNop` in our pystatsd repo.

**Review by tipabu — COMMENTED** [review]
@charz also pointed out elsewhere that `read` _also_ puts to the metrics queue -- given that there isn't much _else_ it does, I guess we need to guard the call to `register_read`, too...

I wonder if it'd be better for us to just plan on sending swift metrics straight to the statsd-exporter, and get rid of the collectd plugin entirely.

**Comment by charz** [comment]
>I wonder if it'd be better for us to just plan on sending swift metrics straight to the statsd-exporter, and get rid of the collectd plugin entirely.

We still need some metrics from collectd (collectd-plugin) to statsd-exporter, but it'll have a bunch of metrics need to fix, and I'd to make this work. And we can move forward
1. change swift metrics to statsd-exporter
2. collect collectd metrics from prometheus plugin or just use node-exporter and other exporters to replace it.

**Pull request merged**